### ISSUE TITLE
Is there any way to clean memory for weights and parameters programatically?

### ISSUE THREAD

**Issue by masakinakada** [fchollet/keras#6325]
Dear community, 

I would like to try different architectures with same datasete, and am wonderfing if there is a way to clean the memory for parameters keeping the training dataset loaded in the system. 

I checked stackover flow and other website, but could not find the solution. 
My dataset is pretty big and takes long time to load, so I want to avoid the data importing everytime. 

Thank you very much in advance.

**Comment by bstriner** [comment]
Just make a new instance of the model with your new architecture. Not sure where you would be running into problems.

Cheers

**Comment by masakinakada** [comment]
Thanks Ben!
The problem I wnted to avoid was the reloading of the dataset when I test
the new architecture.

Seems like it is the only way though....

Masaki

**Comment by bstriner** [comment]
What are you doing that is reloading the dataset? What does when you load the dataset have to do with Keras? Load the dataset, create a model, train it, create a new model, train it, repeat. Or by load do you mean train?

If you're loading your data into a numpy array somewhere, you can use that data multiple times.

If you want to try multiple architectures, you can actually combine them all into a single Keras model so they can train at the same time. This might be a good idea if you have a handful of models and want to train on a lot of data. Is that what you're looking for?

Cheers

**Comment by masakinakada** [comment]
Hi Ben,

Yes,
>>If you want to try multiple architectures, you can actually combine them
all into a single Keras model so they can train at the same time. This
might be a good idea if you have a handful of models and want to train on a
lot of data. Is that what you're looking for?

This is what I was looking for! I did not know that I can use single Keras
model and combine all of the architecture togheter and run train all at the
same time. I would like to check how to do that!

I tried to load the data to numpy array and reused.
This works for the memory management of training dataset, but somehow, the
allocated memory for weights parameters were not deallocated and the memory
for that kept increased. In order to solve the problem, I had to kill the
python process in bash script and run a new python script. If you have
solution for this, please kindly give me advice..
Since I am training pretty deep network, it seems the memory goes always
close to the maximum capacity fro 32GB RAM.

Thank you so much Ben!

Masaki

**Comment by bstriner** [comment]
1) If your models are so big you can only fit one in memory at a time, you will have to do them sequentially. The parameters shouldn't be sitting around in memory. Make sure you remove references. You can try `del` and `gc.collect`. Make sure the models are generated within their own closure so the references don't leak. As long as you do standard python things the garbage should be collected.

Theano garbage collection should be enabled by default so that shouldn't be the issue but you can try enabling it explicitly to be sure.

If you remove all references to the model and still see a memory leak, maybe post code to reproduce and we can go from there.

2) If you have a ton of input data, you should try using fit_generator instead of fit. This allows you to do something like write a loop over files and return batches from those files, or otherwise generate samples on the fly.

It requires more work on your end to use `fit_generator`, but you can write whatever custom code you want for generating each batch in some sort of efficient streaming manner.

3) If you're training a handful of models, there are a lot of ways you can train them at the same time.

Simple example, but doesn't cover all use cases. Say you have 3 models, this will build a single model with one input and 3 targets. Each model will be trained simultaneously and it will report the loss for each of the three models.

```python
models=[m1, m2, m3]
x = Input(...)
targets = [m(x) for m in models]
model = Model([x], targets)
model.compile(...)
model.fit(...) # train all 3 models
...
# use each of the models
m1.save(...)
```

Cheers

**Comment by masakinakada** [comment]
Thank you very much Ben! 
Those are really helpful. I really apprecaite your kind support.

I will try to control the memory explicitly with del and gc.collect 
Since my data is pretty big, it seems we need to go sequential this time. I hope the above method with better memory management would solve the memory issues. 

I would also like to try fit_generator. Sounds like it is more customizable and it might give us more flexibility for the data size. 

I will give you update on hwo things go at the end! 
Again, thank you so much!

**Comment by bstriner** [comment]
When you have a ton of data you should be streaming your data. That applies to any framework out there.

One idea is to shuffle all your data and save it in numpy files. You write a generator that picks a random file, loads it, shuffles it, and yields each of the batches in the file, then repeats with a new file.

Not sure if there is something premade to do that. Might be worth building a generic utility or something. Basically just a few for loops so not the hardest thing to do in custom code. 

If you really want to go deep, you can read about Elephas, which is for distributed Keras.

https://github.com/maxpumperla/elephas

The parameter memory should be cleared as long as you're not holding references somewhere in your code. If they don't, then maybe there is a hanging reference somewhere in Keras we would need to troubleshoot.

Hypothetically, one work-around is to try to set the parameters to empty matrices. There is no theoretical reason you should have to do this if GC is working correctly, but something like this could work:

```python
for w in model.trainable_weights:
  K.set_value(w, [[0]])
```

**Comment by masakinakada** [comment]
Thank you very much Ben!

Now I finished the implementation of the fit_generator and running it. 
I could use all the data I have here in the approach. 

Each epoch takes quite long time, ( around 2 hours) but hope it will make a good convergence.. 
I would like to give you update on how things go! 

Again, thank you so much!
Masaki

**Comment by bstriner** [comment]
GLHF!
